File: EASY/reverse_string.py
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """

        # Using two pointers

        """
        left, right = 0, len(s) - 1

        while left <= right:
            temp = s[left]            
            s[left] = s[right]
            s[right] = temp

            left += 1
            right -= 1

        """
        # using stack
        # Time: O(n) Space(n)

        """
        stack = []
        for c in s:
            stack.append(c)
        i = 0
        while stack:
            s[i] = stack.pop()
            i += 1
        """

        # reversive method (using l, r pointer)
        def reverse(l, r):
            if l < r:
                s[l], s[r] = s[r], s[l]
                reverse(l+1, r-1)
        reverse(0, len(s) - 1)

File: EASY/Easy_Cpp/reverse_string.cpp
class Solution {
public:
    void reverseString1(vector<char>& s) {
        // Using two pointers
        int left = 0, right = s.size() - 1;
        while (left < right){
            char temp = s[left];
            s[left] = s[right];
            s[right] = temp;
            left++;
            right--;
        }
    }

    void reverseString2(vector<char>& s) {
        // Using stack
        stack<char> stk;
        for (char c : s){
            stk.push(c);
        }
        int i = 0;
        while (!stk.empty()) {
            s[i] = stk.top();
            stk.pop();
            i++;
        }
    }

    void reverseString3(vector<char>& s, int left, int right) {
        // Recursive method
        if (left >= right) {
            return;
        }
        char temp = s[left];
        s[left] = s[right];
        s[right] = temp;
        reverseString3(s, left + 1, right - 1);
    }
};
